compiler/noirc_frontend/src/ast/expression.rs

```diff
@@ -87,6 +88,17 @@ impl UnresolvedGeneric {
             UnresolvedGeneric::Variable(ident) | UnresolvedGeneric::Numeric { ident, .. } => ident,
         }
     }
+
+    fn resolve_numeric_kind_type(typ: UnresolvedType) -> Result<Type, UnresolvedTypeData> {
+        use crate::ast::UnresolvedTypeData::{FieldElement, Integer};
+
+        match typ.typ {
+            FieldElement => Ok(Type::FieldElement),
+            Integer(sign, bits) => Ok(Type::Integer(sign, bits)),
+            // Only fields and integers are supported for numeric kinds
+            _ => Err(typ.typ),
```

Comment:
We always want to make a `DefCollectorErrorKind::UnsupportedNumericGenericType` from this, so we should make it directly here and return that instead. I think it's more clear than returning a `UnresolvedTypeData` on error.

Reply:
I kept this as returning an `UnresolvedTypeData` as this function is only looking to resolve the `UnresolvedType`. Then in `kind` I map it to a `DefCollectorErrorKind::UnsupportedNumericGenericType`.

Reply:
I'd prefer it in here since resolving would imply we'd give an error if it fails to resolve as well. Putting it here also means we can avoid cloning the type in `kind` (since `resolve_numeric_kind` wouldn't need an owned type). There's never a step where we care about the `UnresolvedTypeData` returned in the error case so I think we can avoid exposing it.

Reply:
Ok I have changed it to return the def collector error from `resolve_numeric_kind_type`